rollouts/pkg/packagediscovery/packagediscovery.go

```diff
@@ -149,29 +260,69 @@ func (d *PackageDiscovery) getPackagesForRepository(gitHubClient *github.Client,
 		directories := filterByPattern(selector.Directory, allDirectories)
 
 		for _, directory := range directories {
-			thisDiscoveredPackage := DiscoveredPackage{Org: selector.Org, Repo: repoName, Revision: selector.Revision, Directory: directory}
+			thisDiscoveredPackage := DiscoveredPackage{
+				Revision:   selector.Revision,
+				Directory:  directory,
+				GitHubRepo: repo,
+				Branch:     branch,
+			}
 			discoveredPackages = append(discoveredPackages, thisDiscoveredPackage)
 		}
 	} else {
-		thisDiscoveredPackage := DiscoveredPackage{Org: selector.Org, Repo: repoName, Revision: selector.Revision, Directory: selector.Directory}
+		thisDiscoveredPackage := DiscoveredPackage{
+			Revision:   selector.Revision,
+			Directory:  selector.Directory,
+			GitHubRepo: repo,
+			Branch:     branch,
+		}
 		discoveredPackages = append(discoveredPackages, thisDiscoveredPackage)
 	}
 
 	return discoveredPackages, nil
 }
 
-func (d *PackageDiscovery) getGitHubClient(ctx context.Context, selector gitopsv1alpha1.GitHubSelector) (*github.Client, error) {
+func (d *PackageDiscovery) useCache(config gitopsv1alpha1.PackagesConfig) bool {
+	return d.cache != nil && cmp.Equal(config, d.cache.config) && time.Now().Before(d.cache.expiration)
+}
+
+func (d *PackageDiscovery) NewGitLabClient(ctx context.Context, gitlabSource gitopsv1alpha1.GitLabSource) (*gitlab.Client, error) {
+
+	// initialize a gitlab client
+	secretName := gitlabSource.SecretRef.Name
+	if secretName == "" {
+		return nil, fmt.Errorf("GitLab secret reference is missing from the config")
+	}
+	var repositorySecret coreapi.Secret
+	key := client.ObjectKey{
+		Namespace: d.namespace,
+		Name:      secretName,
+	}
+	if err := d.client.Get(ctx, key, &repositorySecret); err != nil {
+		return nil, fmt.Errorf("cannot retrieve gitlab credentials %s: %v", key, err)
+	}
+
+	accessToken := string(repositorySecret.Data["token"])
+	// TODO(droot): BaseURL should also be configurable through the API
```

Comment:
question: what other baseURLs are possible with gitlab?

Reply:
Two use-cases that I have in my mind:

1. Enterprises can self-host gitlab instances and in that case, capability to point to self-hosted gitlab will be needed
2. Eventually, we will use `gitlab` docker instance for our e2e tests where being able to point to our instance will be needed.